rfd/0034-desktop-access-windows.md

```diff
@@ -0,0 +1,256 @@
+exposure to Kerberos tickets (even when using AAD as SSO provider) and we can't
+assume that all clients use Windows machines, Teleport will not use TGTs for
+RDP authentication.
+
+### Host discovery
+
+When a user starts a new desktop session, they must specify the Windows host to
+connect to. Internally, Teleport tracks known Windows hosts using
+`WindowsDesktop` objects.
+
+There are 3 ways that `windows_desktop_service` discovers Windows hosts to
+register:
+- hardcoded list of standalone hosts provided in the config file (see
+  [configuration](#configuration))
+- list of Active Directory-enrolled hosts obtained from AD via LDAPS (LDAP over
+  SSL)
+  - LDAP library: https://pkg.go.dev/github.com/go-ldap/ldap/v3
+- local host, when running on a Windows machine in agent mode
+
+### Concurrent sessions
+
+RDP supports concurrent user sessions on the same host. It allocates a virtual
+desktop for each user to isolate their activities. However, it only allows a
+single session per user per host. If a user starts a new session on the host
+that they are already logged into, they will log out the other session. This is
+RDP behavior we cannot change.
+
+All the usual controls on Teleport sessions apply to Desktop Access, like user
+locking, concurrent session limits and idle timeouts.
+
+### Configuration
+
+New `teleport.yaml` section for `windows_desktop_service`:
+
+```yaml
+windows_desktop_service:
+  enabled: yes # default false
+  # listen_addr can share the port with the proxy web port using SNI.
+  listen_addr: 0.0.0.0:3080
+  public_addrs: [rdp.example.com:3080]
+  mode: "gateway" # or "agent"
+  # (optional) ldap contains hostname and credentials for the LDAP server on
+  # the Active Directory domain controller.
+  # If specified, windows hosts will be automatically discovered by Teleport.
+  #
+  # Note: Teleport will only connect to LDAP over TLS and will always
+  # validate the server certificate.
+  ldap:
+    host: ldap.example.com
+    username: "ldap_user"
+    password: "ldap_pass"
+    # optional CA cert to use for LDAP server certificate validation.
+    ca_cert: "/var/lib/teleport/ldap_ca.pem"
+  # (optional) hosts is a list of hostnames to register as WindowsDesktop
+  # objects in Teleport.
+  # These are usually non-AD hosts.
+  hosts:
+  - win1.example.com
+  - win2.example.com
+  - ...
+  # (optional) host_labels applies labels to windows hosts for RBAC.
+  # Each entry maps to a subset of hosts by regexp and applies a group of labels.
+  # A host can match multiple regexps and will get a union of all the labels.
+  host_labels:
+  - match: ".*"
+    labels:
+    - env: prod
+  - match: "^db.*"
+    labels:
+    - type: database
+  - match: "^dc\.example\.com$"
+    labels:
+    - type: domain_controller
+```
+
+### Security concerns
+
+RDP is a very complex protocol with a history of vulnerabilities. Using a
+memory-safe client that we fully control and not exposing raw RDP publicly is
```

Comment:
This would have to be developed in a memory-safe language since you have to support the full RDP protocol (complex state machine!) and you might be connecting to untrusted RDP services. We should plan for extensive fuzzing on this component alone.

Reply:
Yep, the initial attempt is using a Rust library: https://github.com/gravitational/teleport/pull/7824
To call into it from Go we need some unsafe C glue, which is a good area to focus on during review.